Add "Select on send" option to support both transports functionality

jd-alexander opened the pull request:
Closes #2371

#### What has been done to verify that this works as intended?
`Select on send` was enabled under `Send submissions via` and once this was done both buttons appeared on the upload list. When either the `SMS` or `Internet` button is pressed the respective functionality was triggered. Once you go to the settings screen and switch back to either `Wifi/Cellular` or `SMS` then the extra SMS button disappears and the others resize.

#### Why is this the best possible solution? Were any other approaches considered?
This is the best solution because putting the buttons on the screen make the functionality easy to access. There was a bit of consideration to use a dialog or bottom sheet but that would add an unnecessary extra step.

#### Are there any risks to merging this code? If so, what are they?
No. The only risk is to ensure that regression doesn't take place.

#### Do we need any specific form for testing your changes? If so, please attach one.
A SMS form.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lintDebug spotbugsDebug` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

jd-alexander committed:
Added strings to support both transports functionality

Values were added to support the new SMS upload button and also to power the "Select on send" transport option.

jd-alexander committed:
Modified conditions in ServerPreferencesFragment for the new "Select on send" option

The conditions were only reacting to the internet and sms transports so they were modified to support "Both" where in SMS and Both make the SMS Submission settings enabled.

jd-alexander committed:
Refactored button holder layout for the new button

Added the new SMS upload button that will be visible when Both transports is enabled. The layout was refactored to support all children of the button holder resizing based on the tallest child.

jd-alexander committed:
Implemented Both transports functionality in the InstanceUploaderList

This involved setting up conditions to toggle the visibility of the SMS upload button, modifying the upload button text and also knowing which upload functionality to use depending on the button clicked.

jd-alexander committed:
Upload button now gets the text set when the transport type changes.

jd-alexander committed:
removed blank lines.

jd-alexander commented on collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:
```diff
@@ -160,10 +160,10 @@ public void setupTransportPreferences() {
                         smsGatewayPreference.setEnabled(true);
                         smsPreferenceCategory.setEnabled(true);
-                        transportPreference.setSummary(R.string.transport_type_sms);
+                        transportPreference.setSummary(newValue.equals(getString(R.string.transport_type_value_sms)) ? R.string.transport_type_sms : R.string.transport_type_both);
```
@lognaturel I just saw your ternary comment! Should I pull this summary out in a variable or just use a normal if statement? I just used the ternary because it felt concise for this use case.

lognaturel replied:
I would definitely at least split this into two lines to stay under the 80-char mark and help with readability. Having longer lines makes it harder to skim things, means there's a need for horizontal scrolling on Github and makes it harder to read code from a laptop or with a horizontally-split screen. It also usually suggests there's probably a simpler way.

In this case, I think this clarifies intent:
```
boolean smsEnabled = !newValue.equals(getString(R.string.transport_type_value_internet));
smsGatewayPreference.setEnabled(smsEnabled);
smsPreferenceCategory.setEnabled(smsEnabled);

if (newValue.equals(...) // three-way if to set summary.
```

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java:
```diff
@@ -132,55 +141,74 @@ public void denied() {
+    @OnClick({R.id.upload_button, R.id.sms_upload_button})
+    public void onUploadButtonsClicked(Button button) {
+        String transport = (String) GeneralSharedPreferences.getInstance().get(KEY_SUBMISSION_TRANSPORT_TYPE);
+        if (transport.equalsIgnoreCase(getString(R.string.transport_type_value_internet)) || button.getId() == R.id.upload_button) {
```
I think there are some inconsistencies between `equals` and `equalsIgnoreCase` in different parts of the code. Maybe it's worth introducing a `isInternetTransportSelected`, `isSMSTransportSelected`, etc methods to make sure usage is consistent and reduce the cognitive load of some of these lines.

jd-alexander replied:
That's a good point! Will do that.

grzesiek2010 replied:
Didn't you want to use the enum class you implemented here also? I think you forgot about it.

jd-alexander replied:
Oh yes, thanks.

lognaturel commented:
Once those two small changes discussed inline are made, this can go to testing!

jd-alexander committed:
Introduced an enum called Transport for submission transport types

This helps with reducing the cognitive load it takes to read the string comparisons to differentiate between transport types.

jd-alexander commented:
Ready for testing!

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into both_transports

jd-alexander committed:
Optimized Transport enum implementation

Converted the string comparisons in onUploadButtonsClicked and also did a String cast inside fromPreference to remove the redundant casts.

mmarciniak90 commented:
**Tested with success!**

Verified on Android: 4.1, 4.4, 5.1, 6.0, 7.0, 8.1
Verified cases:
* Send (Internet)
* Send (SMS)
* Change settings from Send Finalized Form page
* only SMS or Wifi/Cellular options
* Minimize app, lock screen

|![screenshot_20180809-123336](https://user-images.githubusercontent.com/3907554/43892471-8d7e0594-9bbb-11e8-9362-47ab73a807b8.png)|![screenshot_20180809-123313](https://user-images.githubusercontent.com/3907554/43892472-8db13aae-9bbb-11e8-8664-d1d2e67ef8c1.png)|
|---|---|

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

mmarciniak90 commented:
@opendatakit-bot label "behavior verified"

grzesiek2010 committed:
Merge branch 'master' into both_transports
